File: impl/typescript/test/finished-kat.test.ts
// Standards-derived, NON-CIRCULAR known-answer test for the draft-00 Finished verify_data
// (binding spec/10 §6.2; RFC 8446 §4.4.4): verify_data = HMAC(finished_key, transcript_hash) under
// the profile hash (SHA-256 at Standard). TypeScript mirror of the Go reference test against
// the SAME pinned vector.
//
// NON-CIRCULARITY: the expected verify_data in test-vectors/v1/finished-kat.json were produced with
// HMAC directly, NOT with computeFinished. Three legs:
//  1. ANCHOR — HMAC-SHA-256 reproduces the published RFC 4231 TC1/TC2 MACs (trust the primitive).
//  2. ORACLE — an independent createHmac reproduces verify_data (guards the vector).
//  3. IMPL   — computeFinished reproduces verify_data; verifyFinished accepts the correct MAC and
//     rejects a single-bit tamper (guards the implementation).
// finished_key inputs are fixtures (their DERIVATION is anchored by the key-schedule KAT); the
// transcript_hash inputs are the Transcript KAT's TH_sCV / TH_cCV (the points §6.2 covers).
import { test } from "node:test";
import assert from "node:assert";
import { createHash, createHmac } from "node:crypto";
import { readFileSync } from "node:fs";
import { join } from "node:path";
import { computeFinished, verifyFinished } from "../src/npamp.ts";

const VECTORS = join(import.meta.dirname, "..", "..", "..", "test-vectors", "v1");
const FINISHED_KAT_SHA256 = "25c21b0bd3b3b6b77862f4a819f81ff5e4ff42e4b1d70af81feeedc5aad73c7f";

const hx = (s) => Buffer.from(s, "hex");

function loadFinishedKAT() {
  const raw = readFileSync(join(VECTORS, "finished-kat.json"));
  const got = createHash("sha256").update(raw).digest("hex");
  assert.equal(got, FINISHED_KAT_SHA256, "Finished KAT vector SHA-256 mismatch (swapped vector?)");
  return JSON.parse(raw.toString("utf8"));
}

// hmacOracle is the standard HMAC-SHA-256, independent of computeFinished.
function hmacOracle(key, data) {
  return createHmac("sha256", key).update(data).digest();
}

// ANCHOR: HMAC-SHA-256 reproduces the published RFC 4231 vectors before any verify_data is trusted.
test("finished_kat_rfc4231_anchor", () => {
  const k = loadFinishedKAT();
  for (const [name, tc] of [["TC1", k.rfc4231_hmac_sha256.tc1], ["TC2", k.rfc4231_hmac_sha256.tc2]]) {
    assert.equal(hmacOracle(hx(tc.key), hx(tc.data)).toString("hex"), tc.hmac_sha256, `HMAC-SHA-256 ${name} != RFC 4231`);
  }
});

// ORACLE: reproduce verify_data with an independent createHmac (guards the vector).
test("finished_kat_oracle", () => {
  const k = loadFinishedKAT();
  const n = k.npamp_inputs, e = k.expected;
  assert.equal(hmacOracle(hx(n.finished_key_server), hx(n.th_scv)).toString("hex"), e.verify_data_server, "oracle server verify_data");
  assert.equal(hmacOracle(hx(n.finished_key_client), hx(n.th_ccv)).toString("hex"), e.verify_data_client, "oracle client verify_data");
});

// IMPL: computeFinished reproduces verify_data; verifyFinished accepts the correct MAC, rejects a tamper.
test("finished_kat_impl", () => {
  const k = loadFinishedKAT();
  const n = k.npamp_inputs, e = k.expected;
  for (const c of [
    { name: "server", fk: n.finished_key_server, th: n.th_scv, want: e.verify_data_server },
    { name: "client", fk: n.finished_key_client, th: n.th_ccv, want: e.verify_data_client },
  ]) {
    const fk = hx(c.fk), th = hx(c.th), want = hx(c.want);
    assert.equal(computeFinished(fk, th, true).toString("hex"), c.want, `[${c.name}] computeFinished mismatch`);
    assert.ok(verifyFinished(fk, th, want, true), `[${c.name}] verifyFinished rejected the correct verify_data`);
    const bad = Buffer.from(want);
    bad[0] ^= 0x01;
    assert.ok(!verifyFinished(fk, th, bad, true), `[${c.name}] verifyFinished accepted a tampered verify_data`);
  }
});

File: impl/go/finished_kat_test.go
package npamp

import (
	"bytes"
	"crypto/hmac"
	"crypto/sha256"
	"errors"
	"testing"
)

// finishedKAT mirrors test-vectors/v1/finished-kat.json (ADR-0010):
// verify_data = HMAC-SHA256(finished_key, transcript_hash) with the HMAC
// primitive anchored to RFC 4231 TC1/TC2, fixed finished_key fixtures, and the
// Transcript KAT's TH_sCV / TH_cCV as the covered points (spec/10 section
// 6.2).
type finishedKAT struct {
	Expected struct {
		Client string `json:"verify_data_client"`
		Server string `json:"verify_data_server"`
	} `json:"expected"`
	Inputs struct {
		FinishedKeyClient string `json:"finished_key_client"`
		FinishedKeyServer string `json:"finished_key_server"`
		THCCV             string `json:"th_ccv"`
		THSCV             string `json:"th_scv"`
	} `json:"npamp_inputs"`
	RFC4231 struct {
		TC1 struct {
			Data string `json:"data"`
			MAC  string `json:"hmac_sha256"`
			Key  string `json:"key"`
		} `json:"tc1"`
		TC2 struct {
			Data string `json:"data"`
			MAC  string `json:"hmac_sha256"`
			Key  string `json:"key"`
		} `json:"tc2"`
	} `json:"rfc4231_hmac_sha256"`
}

// hmacSHA256 is the test's oracle: crypto/hmac applied directly, independent
// of ComputeFinished's wiring.
func hmacSHA256(key, data []byte) []byte {
	m := hmac.New(sha256.New, key)
	m.Write(data)
	return m.Sum(nil)
}

func TestFinishedKAT(t *testing.T) {
	var kat finishedKAT
	loadKAT(t, "finished-kat.json", &kat)

	// Anchor (RFC 4231 TC1/TC2): prove the HMAC-SHA-256 primitive.
	for name, tc := range map[string]struct{ Data, MAC, Key string }{
		"tc1": {kat.RFC4231.TC1.Data, kat.RFC4231.TC1.MAC, kat.RFC4231.TC1.Key},
		"tc2": {kat.RFC4231.TC2.Data, kat.RFC4231.TC2.MAC, kat.RFC4231.TC2.Key},
	} {
		got := hmacSHA256(mustHex(t, name+".key", tc.Key), mustHex(t, name+".data", tc.Data))
		if !bytes.Equal(got, mustHex(t, name+".hmac", tc.MAC)) {
			t.Fatalf("HMAC-SHA-256 does not reproduce RFC 4231 %s", name)
		}
	}

	cases := []struct {
		role        string
		finishedKey []byte
		th          []byte
		want        []byte
	}{
		{"server", mustHex(t, "finished_key_server", kat.Inputs.FinishedKeyServer),
			mustHex(t, "th_scv", kat.Inputs.THSCV), mustHex(t, "verify_data_server", kat.Expected.Server)},
		{"client", mustHex(t, "finished_key_client", kat.Inputs.FinishedKeyClient),
			mustHex(t, "th_ccv", kat.Inputs.THCCV), mustHex(t, "verify_data_client", kat.Expected.Client)},
	}
	for _, c := range cases {
		// Oracle leg: independent crypto/hmac reproduces the vector.
		if !bytes.Equal(hmacSHA256(c.finishedKey, c.th), c.want) {
			t.Fatalf("%s: oracle HMAC does not reproduce the vector", c.role)
		}
		// Impl leg: ComputeFinished reproduces it through the real code path.
		if got := ComputeFinished(c.finishedKey, c.th, ProfileStandard); !bytes.Equal(got, c.want) {
			t.Fatalf("%s: ComputeFinished = %x, want %x", c.role, got, c.want)
		}
		// VerifyFinished accepts the correct MAC ...
		if err := VerifyFinished(c.finishedKey, c.th, c.want, ProfileStandard); err != nil {
			t.Fatalf("%s: VerifyFinished rejected the correct verify_data: %v", c.role, err)
		}
		// ... and rejects a corrupted one and a wrong-key one.
		bad := bytes.Clone(c.want)
		bad[0] ^= 0x01
		if err := VerifyFinished(c.finishedKey, c.th, bad, ProfileStandard); !errors.Is(err, ErrFinishedMismatch) {
			t.Fatalf("%s: corrupted verify_data not rejected (err=%v)", c.role, err)
		}
		wrongKey := bytes.Clone(c.finishedKey)
		wrongKey[0] ^= 0x01
		if err := VerifyFinished(wrongKey, c.th, c.want, ProfileStandard); !errors.Is(err, ErrFinishedMismatch) {
			t.Fatalf("%s: wrong finished_key not rejected (err=%v)", c.role, err)
		}
	}

	// Cross-direction guard: the server MAC keyed with the client key (or over
	// the client transcript point) MUST NOT verify.
	if err := VerifyFinished(
		mustHex(t, "finished_key_client", kat.Inputs.FinishedKeyClient),
		mustHex(t, "th_scv", kat.Inputs.THSCV),
		mustHex(t, "verify_data_server", kat.Expected.Server),
		ProfileStandard,
	); !errors.Is(err, ErrFinishedMismatch) {
		t.Fatalf("server verify_data verified under the client finished_key (err=%v)", err)
	}
}
